lib/puppet/provider/mysql_grant/mysql.rb

```diff
@@ -129,11 +131,28 @@ def flush
 
   mk_resource_methods
 
+  def diff_privileges(privileges_old, privileges_new)
+    diff = {:revoke => Array.new, :grant => Array.new}
+    if privileges_new.include? 'ALL' or privileges_old.include? 'ALL'
+      diff[:revoke] = privileges_old
+      diff[:grant] = privileges_new
+    else
+      diff[:revoke] = privileges_old - privileges_new
+      diff[:grant] = privileges_new - privileges_old
+    end
+    return diff
+  end
+
   def privileges=(privileges)
-    revoke(@property_hash[:user], @property_hash[:table])
-    grant(@property_hash[:user], @property_hash[:table], privileges, @property_hash[:options])
+    diff = diff_privileges(@property_hash[:privileges], privileges)
+    if not diff[:revoke].empty?
+      revoke(@property_hash[:user], @property_hash[:table], diff[:revoke])
+    end
+    if not diff[:grant].empty?
+      grant(@property_hash[:user], @property_hash[:table], diff[:grant], @property_hash[:options])
+    end
     @property_hash[:privileges] = privileges
-
+    mysql([defaults_file, '-NBe', 'FLUSH PRIVILEGES'].compact)
```

Comment:
Flush privileges is already being done in the `flush` method, so we don't need it here.